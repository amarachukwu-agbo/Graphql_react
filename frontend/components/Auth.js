import React from 'react';
import { Query } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';
import SignIn from '../components/SignIn';

const Auth = ({ children }) => {
  return (
    <Query query={CURRENT_USER_QUERY}>
      {({ data , loading}) => {
        if (loading) return <p>Loading ...</p>
        if (data && !data.me) return <SignIn />
        return children;
      }}
    </Query>
  );
};

export default Auth;