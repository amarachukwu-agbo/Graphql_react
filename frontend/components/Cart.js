import React, { Component } from 'react';
import gql from 'graphql-tag';
import StripeCheckout from 'react-stripe-checkout';
import Router from 'next/router';
import NProgress from 'nprogress';
import { Query, Mutation } from 'react-apollo';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';
import User, { CURRENT_USER_QUERY } from './User';
import CartItem from './CartItem';
import calculateTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';

export const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;

export const CREATE_ORDER = gql`
  mutation CREATE_ORDER($token: String!) {
    createOrder(token: $token) {
      id
      items {
        image
        description
        title
      }
      total
      charge
    }
  }
`;

const totalItems = (cart) => {
  return cart.reduce((tally, item) => tally + item.quantity, 0)
}
class Cart extends Component {
  onToken = async (token, createOrder) => {
    NProgress.start();
    const order = await createOrder({
      variables: {
        token: token.id
      }
    }).catch(error => alert(`Error: ${error}`));
    Router.push({
      pathname: '/order',
      query: { id: order.data.createOrder.id }
    });
  }
  render() {
    return (
      <User>
        {({ data } ) => {
          const { me } = data;
          if (!me) return null;
          return(
            <Mutation mutation={TOGGLE_CART_MUTATION}>
              {(toggleCart) => (
              <Query query={LOCAL_STATE_QUERY}>
                {({ data }) => (
                  <CartStyles open={data.cartOpen}>
                    <header>
                      <CloseButton title="close" onClick={toggleCart}>&times;</CloseButton>
                      <Supreme>{me.name}'s Cart</Supreme>
                      <p>You have {me.cart.length} item{me.cart.length === 1 ? '': 's'} in your cart</p>
                    </header>
                    <ul>
                      { me.cart.map(item => <CartItem key={item.id} cartItem={item}/>) }
                    </ul>
                    <footer>
                      {
                        me.cart.length ? (
                          <>
                            <p>{formatMoney(calculateTotalPrice(me.cart))}</p>
                            <Mutation mutation={CREATE_ORDER} refetchQueries={[{query: CURRENT_USER_QUERY }]}>
                              {(createOrder) => (
                                <StripeCheckout
                                  name="Sick Fits"
                                  description={`Order of ${totalItems(me.cart)} items`}
                                  token={(stripeToken) => this.onToken(stripeToken, createOrder)}
                                  stripeKey="pk_test_4ApAJihRk5yShkglGcJ0ZSXV00oEwKYLu4"
                                  amount={calculateTotalPrice(me.cart)}
                                  image={me.cart.length && me.cart[0].item && me.cart[0].item.image}
                                  currency="USD"
                                >
                                  <SickButton>Checkout</SickButton>
                                </StripeCheckout>
                              )}
                            </Mutation>
                          </>
                        ) : ''
                      }
                    </footer>
                  </CartStyles>
                )}
              </Query>
              )}
            </Mutation>
          )
        }}
      </User>
    );

  }
};

export default Cart;