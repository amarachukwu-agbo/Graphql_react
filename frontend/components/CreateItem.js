import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import Router from 'next/router';

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;
export default class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  }

  handleChange = (e) => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({
      [name]: val
    });
  }

  uploadImage = async (e) => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');
    const res = await fetch('https://api.cloudinary.com/v1_1/ama-hello-books-v2/image/upload', {
      method: 'POST',
      mode: 'cors',
      body: data
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    });
  }
  render() {
    const { title, price, description, image } = this.state;
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
      {(createItem, { loading, error}) => (
        <Form onSubmit={async (e) => {
          e.preventDefault();
          const res = await createItem();
          Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id}
          });
        }}>
          <h2> Sell an item</h2>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
          <label htmlFor="file">
            File
            </label>
            <input
              type="file"
              id="file"
              name="file"
              placeholder="Upload an image"
              required
              onChange={this.uploadImage}
            />
            {
              image && <img src={image} alt="uploaded image" width="50px"/>
            }
            <label htmlFor="title">
            Title
            </label>
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Title"
              required
              value={title}
              onChange={this.handleChange}
            />

            <label htmlFor="price">
            Price
            </label>
            <input
              type="number"
              id="price"
              name="price"
              placeholder="Price"
              required
              value={price}
              onChange={this.handleChange}
            />

            <label htmlFor="description">
            Description
            </label>
            <textarea
              type="text"
              id="description"
              name="description"
              placeholder="Enter a description"
              required
              value={description}
              onChange={this.handleChange}
            />
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
      </Mutation>
    )
  }
}
