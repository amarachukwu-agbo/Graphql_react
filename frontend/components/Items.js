import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Item from './Item';
import Pagination from './Pagination';
import { perPage } from '../config';


export const ITEMS_QUERY = gql`
  query ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}){
    items(skip: $skip, first: $first) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  margin: 0 auto;
  max-width: ${props => props.theme.maxWidth};
`;

export default class Items extends Component {
  render() {
    const { page } = this.props;
    return (
      <Center>
      <Query query={ITEMS_QUERY} variables={{
        skip: (page - 1) * perPage
      }}>
        {({ error, loading, data }) => {
          if (loading) return <p>Loading... </p>
          if (error) return <p>{`Error: ${error.message}`}</p>
          console.log(data, 'data =====');
          return data.items.length ? (
            <div>
              <Pagination page={page}/>
              <ItemsList>
                {
                  data.items.map(item => <Item key ={item.id} item={item}/>)
                }
              </ItemsList>
              <Pagination page={page}/>
            </div>
          ) : <p>No items have been created :(</p>
        }}
      </Query>
      </Center>
    )
  }
}
