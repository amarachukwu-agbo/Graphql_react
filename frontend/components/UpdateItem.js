import React, { Component } from 'react';
import { Mutation, Query } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';


export const FETCH_ITEM_QUERY = gql`
  query FETCH_ITEM_QUERY($id: ID!) {
    item(where: {
      id: $id
    }) {
      id
      title
      description
      price
      largeImage
    }
  }
`;

const UPDATE_ITEM_MUTATION = gql`
  mutation UPDATE_ITEM_MUTATION(
    $id: ID!
    $title: String
    $description: String
    $price: Int
    $image: String
    $largeImage: String) {
    updateItem(
      data: {
        title: $title
        description: $description
        price: $price
        image: $image
        largeImage: $largeImage
      }
      where: {
        id: $id
      }
    ) {
      id
    }
  }
`;
export default class UpdateItem extends Component {
  state = {}

  handleChange = (e) => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({
      [name]: val
    });
  }

  updateItem = async (e, updateItemMutation) => {
    e.preventDefault();
    await updateItemMutation();
  }

  render() {
    const { id } = this.props;

    return (
      <Query query={FETCH_ITEM_QUERY} variables={{id}}>
      {({ data, loading }) => {
      if (loading) return <p>Loading ...</p>
      if (!data.item) return <p>{`No item found for id: ${id}`}</p>
      return (
      <Mutation mutation={UPDATE_ITEM_MUTATION} variables={{id, ...this.state}}>
      {(updateItem, { loading, error}) => {
        return (
        <Form onSubmit={async (e) => this.updateItem(e, updateItem)}>
          <h2> Sell an item</h2>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
            <label htmlFor="title">
            Title
            </label>
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Title"
              required
              defaultValue={data.item.title}
              onChange={this.handleChange}
            />

            <label htmlFor="price">
            Price
            </label>
            <input
              type="number"
              id="price"
              name="price"
              placeholder="Price"
              required
              defaultValue={data.item.price}
              onChange={this.handleChange}
            />

            <label htmlFor="description">
            Description
            </label>
            <textarea
              type="text"
              id="description"
              name="description"
              placeholder="Enter a description"
              required
              defaultValue={data.item.description}
              onChange={this.handleChange}
            />
            <button type="submit">Sav{loading ? 'ing' : 'e'} Changes</button>
          </fieldset>
        </Form>
      )}}
      </Mutation>
      )
      }}
      </Query>
    )
  }
}
