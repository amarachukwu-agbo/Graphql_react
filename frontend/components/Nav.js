import Link from 'next/link';
import { Mutation } from 'react-apollo';
import NavStyles from './styles/NavStyles';
import User from '../components/User';
import SignOut from '../components/SignOut';
import CartCount from '../components/CartCount';
import { TOGGLE_CART_MUTATION } from '../components/Cart';

const Nav = () => (
  <Mutation mutation={TOGGLE_CART_MUTATION}>
    {(toggleCart) => (
      <User>
        {({ data: { me }}) => (
          <NavStyles>
            <Link href="/allitems">
              <a>Shop</a>
            </Link>
            {
              me && (
                <>
                  <Link href="/sell">
                    <a>Sell!</a>
                  </Link>
                  <Link href="/orders">
                    <a>Orders</a>
                  </Link>
                  <Link href="/me">
                    <a>Account</a>
                  </Link>
                  <SignOut />
                  <button onClick={toggleCart}>
                    My Cart
                    <CartCount
                      count={me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)}
                    />
                  </button>
                </>
              )
            }
            {
              !me && (
                <Link href="/signup">
                  <a>Sign In</a>
                </Link>
              )
            }
      </NavStyles>
        )}
      </User>
    )}
  </Mutation>
);

export default Nav;
