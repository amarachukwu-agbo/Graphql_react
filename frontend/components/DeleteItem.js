import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
      title
    }
  }
`;

export default class DeleteItem extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired
  }
  update(cache, payload) {
    const { items } = cache.readQuery({ query: ITEMS_QUERY });
    const requiredItems = items
      .filter(item => item.id !== payload.data.deleteItem.id)
    cache.writeQuery({ query: ITEMS_QUERY, data: { items: requiredItems } })
  }
  render() {
    const { id, children } = this.props;
    return (
      <Mutation mutation={DELETE_ITEM_MUTATION}
        variables={{id}}
        update={this.update}
        >
      {(deleteItem) => {
        return (
          <button onClick={() => {
            if (confirm('Are you sure you want to delete this item')) {
              deleteItem().catch(err => alert(err));
            }
          }}>
            { children }
          </button>
        )
      }}
      </Mutation>
    )
  }
}
