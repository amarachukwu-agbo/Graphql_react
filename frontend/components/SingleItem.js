import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import Head from 'next/head';
import styled from 'styled-components';
import { FETCH_ITEM_QUERY } from '../components/UpdateItem';
import ErrorMessage from '../components/ErrorMessage'

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
  .details {
    margin: 3em;
    font-size: 2rem;
  }
`;
export default class SingleItem extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  }

  render() {
    const { id } = this.props;
    return (
      <div>
        <Query query={FETCH_ITEM_QUERY} variables={{id}}>
          {({ error, data: { item }, loading }) => {
            if (error) return <ErrorMessage error={error}/>
            if (loading) return <p>Loading...</p>
            if (!item) return <p>No data found for {`${id}`}</p>
            return (
              <SingleItemStyles>
                <Head>
                  <title> Sick Fits | {item.title} </title>
                </Head>
                <img src={item.largeImage} alt={item.title}/>
                <div className="details">
                  <h2>Viewing {item.title}</h2>
                  <p>{item.description}</p>
                </div>
              </SingleItemStyles>
            )
          }}
        </Query>
      </div>
    )
  }
}
