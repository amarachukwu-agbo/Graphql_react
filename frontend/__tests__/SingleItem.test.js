import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import wait from 'waait';
import { fakeItem } from '../lib/testUtils';
import { FETCH_ITEM_QUERY } from '../components/UpdateItem';
import SingleItem from '../components/SingleItem';

const mocks = [{
  request: {
    query: FETCH_ITEM_QUERY,
    variables: {
      id: '123'
    }
  },
  result: {
    data: {
      item: fakeItem()
    }
  }
}]

describe('<SingleItem />', () => {
  it('renders correctly', () => {
    mount(<MockedProvider mocks={mocks}>
      <SingleItem id="123"/>
    </MockedProvider>);
  });

  it('renders loading when the data is still being fetched', () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <SingleItem id="123"/>
    </MockedProvider>);
    expect(wrapper.find('p').text()).toBe('Loading...');
  });

  it('renders the data when it is fetched', async () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <SingleItem id="123"/>
    </MockedProvider>);
    expect(wrapper.find('p').text()).toBe('Loading...');
    await wait(0);
    wrapper.update();
    expect(wrapper.find('h2')).toMatchSnapshot();
    expect(wrapper.find('img')).toMatchSnapshot();
    expect(wrapper.find('p')).toMatchSnapshot();
  });

  it('renders error when error is returned', async () => {
    const newMocks = [{...mocks[0], result: {errors: [{message:  'No item found'}]}}];
    const wrapper = mount(<MockedProvider mocks={newMocks}>
      <SingleItem id="123"/>
    </MockedProvider>);
    expect(wrapper.find('p').text()).toBe('Loading...');
    await wait(0);
    wrapper.update();
    expect(wrapper.find('[data-test]').text()).toBe('Shoot!No item found');
  });
});
