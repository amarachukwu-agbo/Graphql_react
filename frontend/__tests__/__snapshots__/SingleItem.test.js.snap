// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> renders the data when it is fetched 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleItem /> renders the data when it is fetched 2`] = `
<img
  alt="dogs are best"
  src="dog-small.jpg"
/>
`;

exports[`<SingleItem /> renders the data when it is fetched 3`] = `
<p>
  dogs
</p>
`;
